src/Microsoft.ML.TimeSeries/SrCnnTransformBase.cs

```diff
@@ -0,0 +1,318 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Threading;
+using Microsoft.ML.CommandLine;
+using Microsoft.ML.Data;
+using Microsoft.ML.Internal.Utilities;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.Transforms.TimeSeries
+{
+    internal abstract class SrCnnArgumentBase
+    {
+        [Argument(ArgumentType.Required, HelpText = "The name of the source column.", ShortName = "src",
+            SortOrder = 1, Purpose = SpecialPurpose.ColumnName)]
+        public string Source;
+
+        [Argument(ArgumentType.Required, HelpText = "The name of the new column.",
+            SortOrder = 2)]
+        public string Name;
+
+        [Argument(ArgumentType.AtMostOnce, HelpText = "The size of the sliding window for computing spectral residual", ShortName = "wnd",
+            SortOrder = 3)]
+        public int WindowSize = 24;
+
+        [Argument(ArgumentType.AtMostOnce, HelpText = "The size of the initial window for computingd. The default value is set to 0, which means there is no initial window considered.", ShortName = "iwnd",
+            SortOrder = 4)]
+        public int InitialWindowSize = 0;
+
+        [Argument(ArgumentType.AtMostOnce, HelpText = "The number of points to the back of training window.",
+            ShortName = "backwnd", SortOrder = 5)]
+        public int BackAddWindowSize = 5;
+
+        [Argument(ArgumentType.AtMostOnce, HelpText = "The number of pervious points used in prediction.",
+            ShortName = "aheadwnd", SortOrder = 6)]
+        public int LookaheadWindowSize = 5;
+
+        [Argument(ArgumentType.Required, HelpText = "The size of sliding window to generate a saliency map for the series.",
+            ShortName = "avgwnd", SortOrder = 7)]
+        public int AvergingWindowSize = 3;
+
+        [Argument(ArgumentType.Required, HelpText = "The size of sliding window to generate a saliency map for the series.",
+            ShortName = "jdgwnd", SortOrder = 8)]
+        public int JudgementWindowSize = 21;
+
+        [Argument(ArgumentType.AtMostOnce, HelpText = "The threshold to determine anomaly, score larger than the threshold is considered as anomaly.",
+            ShortName = "thre", SortOrder = 9)]
+        public double Threshold = 0.3;
+    }
+
+    internal abstract class SrCnnTransformBase<TInput, TState> : SequentialTransformerBase<TInput, VBuffer<Double>, TState>
+        where TState : SrCnnTransformBase<TInput, TState>.SrCnnStateBase, new()
+    {
+        internal int BackAddWindowSize;
+
```

Comment:
Are they read-only? If yes, maybe we can append `{ get; }`?

Reply:
Done.